### PR TITLE
Add RobustScanner rec model

### PR DESCRIPTION
**Pull request by tonytonglt**
Thank you for your contribution to the MindOCR repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: mindocr/data/rec_lmdb_dataset.py [hunk 1]
```diff
@@ -186,10 +186,18 @@ def __getitem__(self, idx):
         lmdb_idx, file_idx = self.data_idx_order_list[idx]
         sample_info = self.get_lmdb_sample_info(self.lmdb_sets[int(lmdb_idx)]["txn"], int(file_idx))
 
+        if sample_info is None:
```
Comment (zhtmike) [inline mindocr/data/rec_lmdb_dataset.py:189]: 这两个修改要确保只能在训练集有效，验证集可能会因为这个导致精度误差。另外最好加上warning

Comment (tonytonglt) [reply mindocr/data/rec_lmdb_dataset.py]: 已修改

File: configs/rec/robustscanner/README_CN.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    data_dir: evaluation/                                             # 验证或评估数据集目录，将与`dataset_root`拼接形成完整验证或评估数据集目录
+  ...
+  loader:
+      shuffle: False
+      batch_size: 64                                                  # 验证或评估批大小
+...
+```
+
+**注意:**
+- 由于全局批大小 （batch_size * num_devices） 是对结果复现很重要，因此当GPU/NPU卡数发生变化时，调整`batch_size`以保持全局批大小不变，或将学习率线性调整为新的全局批大小。
```
Comment (HaoyangLee) [inline configs/rec/robustscanner/README_CN.md:295]:
```suggestion
- 由于全局批大小 （batch_size * num_devices） 是对结果复现很重要，因此当GPU/NPU卡数发生变化时，调整`batch_size`以保持全局批大小不变，或根据新的全局批大小线性调整学习率。
```

Comment (HaoyangLee) [reply configs/rec/robustscanner/README_CN.md:295]: 可以同步改下其他rec模型的readme

File: configs/rec/robustscanner/README_CN.md [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+- `en_dict.txt` 是一个包含94个字符的英文字典，其中有数字，常用符号以及大小写的英文字母。
+- `ch_dict.txt` 是一个包含6623个字符的中文字典，其中有常用的繁简体中文，数字，常用符号以及大小写的英文字母。
+
+
+### 自定义词典
+
+您也可以自定义一个字典文件 (***.txt)， 放在 `mindocr/utils/dict/` 下，词典文件格式应为每行一个字符的.txt 文件。
+
+
+如需使用指定的词典，请将参数 `character_dict_path` 设置为字典的路径，并将参数`model->head->out_channels`改为字典中字符的数量+3，`model->head->start_idx`改为字典中字符的数量+1，`model->head->padding_idx`改为字典中字符的数量+2，`loss->ignore_index`改为字典中字符的数量+2。
```
Comment (HaoyangLee) [inline configs/rec/robustscanner/README_CN.md:354]: 把这段描述改成yaml文件的结构格式，然后加以注释说明要改哪里，怎么该。可参考3.1.4章节

File: configs/rec/robustscanner/README_CN.md [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+│   │   └── lock.mdb
+│   ├── ...
+```
+
+然后你可以通过如下修改配置yaml来评估每个数据集，并执行脚本`tools/benchmarking/multi_dataset_eval.py`。
+
+```yaml
+...
+train:
+  # NO NEED TO CHANGE ANYTHING IN TRAIN SINCE IT IS NOT USED
```
Comment (HaoyangLee) [inline configs/rec/robustscanner/README_CN.md:247]:
```suggestion
  # 无需修改训练部分的配置，因验证或推理的时候不必使用该部分
```

File: configs/rec/robustscanner/robustscanner_resnet31.yaml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  ckpt_save_dir: './tmp_rec'
+  dataset_sink_mode: False
+  dataset:
+    type: LMDBDataset
+    dataset_root: path/to/data/ # Optional, if set, dataset_root will be used as a prefix for data_dir
+    data_dir: training/
+    # label_files: # not required when using LMDBDataset
+    sample_ratio: 1.0
+    shuffle: True
+    random_choice_if_none: True
```
Comment (HaoyangLee) [inline configs/rec/robustscanner/robustscanner_resnet31.yaml:80]: 在random_choice_if_none参数后加一个注释说明

File: mindocr/models/backbones/rec_resnet.py [hunk 6]
```diff
@@ -157,6 +157,181 @@ def construct(self, x):
         return [y]
 
 
+class BasicBlockResNet31(nn.Cell):
```
Comment (HaoyangLee) [inline mindocr/models/backbones/rec_resnet.py:160]: 建议把该class放到`class RecResNet(nn.Cell)`上面，让该文件保持Block，@register_backbone_class，@register_backbone的结构

File: configs/rec/robustscanner/README_CN.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+### 自定义词典
+
+您也可以自定义一个字典文件 (***.txt)， 放在 `mindocr/utils/dict/` 下，词典文件格式应为每行一个字符的.txt 文件。
+
+
+如需使用指定的词典，请将参数 `character_dict_path` 设置为字典的路径，并将参数`model->head->out_channels`改为字典中字符的数量+3，`model->head->start_idx`改为字典中字符的数量+1，`model->head->padding_idx`改为字典中字符的数量+2，`loss->ignore_index`改为字典中字符的数量+2。
+
+
+**注意：**
```
Comment (HaoyangLee) [inline configs/rec/robustscanner/README_CN.md:357]: 同上

File: mindocr/models/backbones/rec_resnet.py [hunk 8]
```diff
@@ -157,6 +157,181 @@ def construct(self, x):
...
+            out_channel,
+            kernel_size=3,
+            stride=stride,
+            pad_mode="pad",
+            padding=1)
+
+
+@register_backbone_class
+class RecResNet31(nn.Cell):
+    '''
```
Comment (zhtmike) [inline mindocr/models/backbones/rec_resnet.py:213]: `mindocr/models/backbones/rec_master.py` 内已有resnet31，是否可以合并？

File: mindocr/losses/rec_loss.py [hunk 9]
```diff
@@ -64,8 +64,28 @@ def __init__(self, reduction: str = "mean", ignore_index: int = 0) -> None:
...
-        labels = labels[:, 1:]  # wihout <GO> symbol
+        labels = labels[:, 1:]  # without <GO> symbol
         num_classes = logits.shape[-1]
         logits = ops.reshape(logits, (-1, num_classes))
         labels = ops.reshape(labels, (-1,))
         return self.criterion(logits, labels)
+
+
+class SARLoss(nn.Cell):
+    def __init__(self, **kwargs):
```
Comment (zhtmike) [inline mindocr/losses/rec_loss.py:75]: SARLoss 和 AttentionLoss 非常相似，是否可以合并？

File: configs/rec/robustscanner/README_CN.md [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+|:-------------:|:--------------:|:---------:|:---------:|:---------------------:|:-------:|:----------------------------------------------------------------------------------------------------------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
+| RobustScanner | D910x4-MS2.0-G | ResNet-31 |  87.51%   |     22560 s/epoch     |   310   | [yaml](https://github.com/mindspore-lab/mindocr/blob/main/configs/rec/robustscanner/robustscanner_resnet31.yaml) | [ckpt](https://download.mindspore.cn/toolkits/mindocr/robustscanner/robustscanner_resnet31-72710b5f.ckpt) \| [mindir](https://download.mindspore.cn/toolkits/mindocr/robustscanner/robustscanner_resnet31-72710b5f-72896d92.mindir) |
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md]: 已补充

File: configs/rec/robustscanner/README_CN.md [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+- RobustScanner的MindIR导出时的输入Shape均为(1, 3, 48, 160)。
+
+## 3. 快速开始
+### 3.1 环境及数据准备
+
+#### 3.1.1 安装
+环境安装教程请参考MindOCR的 [installation instruction](https://github.com/mindspore-lab/mindocr#installation).
+
+#### 3.1.2 数据集下载
+本作RobustScanner训练、验证使用的数据集参考了mmocr和PaddleOCR所使用的数据集对文献算法进行复现，在此非常感谢mmocr和PaddleOCR，提高了本repo的复现效率。
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md]: 已修正

File: configs/rec/robustscanner/README_CN.md (hunk 2 above)
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md]: 已修改

File: configs/rec/robustscanner/README_CN.md (hunk 3 above)
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md]: 已修改

File: configs/rec/robustscanner/README_CN.md (hunk 4 above)
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md]: 已修改

File: configs/rec/robustscanner/robustscanner_resnet31.yaml (hunk 5 above)
Comment (tonytonglt) [reply configs/rec/robustscanner/robustscanner_resnet31.yaml]: 已添加

File: mindocr/models/backbones/rec_resnet.py (hunk 6 above)
Comment (tonytonglt) [reply mindocr/models/backbones/rec_resnet.py]: 已修改

File: configs/rec/robustscanner/README_CN.md (hunk 7 above)
Comment (tonytonglt) [reply configs/rec/robustscanner/README_CN.md:385]: 已修改

File: mindocr/data/transforms/rec_transforms.py [hunk 12]
```diff
@@ -569,3 +571,234 @@ def __call__(self, data):
         data["label"] = label
 
         return data
+
+
+class SARLabelEncode(object):
+    """Convert between text-label and text-index"""
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/data/transforms/rec_transforms.py:577]: 已查看，SARLabelEncode 和 RecAttnLabelEncode 对特殊情况的返回None的处理不同，使用的special_char不同，不建议合并

File: mindocr/data/transforms/rec_transforms.py [hunk 13]
```diff
@@ -569,3 +571,234 @@ def __call__(self, data):
...
+    resized_image /= 0.5
+    resize_shape = resized_image.shape
+    padding_im = -1.0 * np.ones((imgC, imgH, imgW_max), dtype=np.float32)
+    padding_im[:, :, 0:resize_w] = resized_image
+    pad_shape = padding_im.shape
+
+    return padding_im, resize_shape, pad_shape, valid_ratio
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/data/transforms/rec_transforms.py]: 已删除

File: mindocr/utils/dict/dict90.txt [hunk 14]
```diff
@@ -0,0 +1,90 @@
+0
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/utils/dict/dict90.txt:1]: 确认为相同字典，已删除

File: mindocr/postprocess/rec_postprocess.py [hunk 15]
```diff
@@ -349,3 +350,114 @@ def __init__(
...
+            return {"texts": text}
+        label = self.decode(label, is_remove_duplicate=False)
+        pred = {"texts": text, "labels": label}
+        return pred
+
+    def get_ignored_tokens(self):
+        return [self.padding_idx]
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (tonytonglt) [reply mindocr/postprocess/rec_postprocess.py]: 已删除

File: mindocr/losses/rec_loss.py (hunk 9 above)
Comment (tonytonglt) [reply mindocr/losses/rec_loss.py:75]: 两者使用的loss的计算均为CrossEntropyLoss，但模型输出的pred shape不同，处理方式不同，不建议合并

**Review by zhtmike — APPROVED** [review]

File: mindocr/models/backbones/rec_resnet.py (hunk 8 above)
Comment (tonytonglt) [reply mindocr/models/backbones/rec_resnet.py:213]: 已查看，网络结构基本一致，但细节上有细微区别，如maxpool2d参数、downsample、channel数量等，我会跑一次实验看精度是否会下降，如果不下降可以进行复用，但可能后续需要将resnet31提取为公共模块

Comment (zhtmike) [reply mindocr/models/backbones/rec_resnet.py:213]: 嗯如果可以复用可以把resnet31搬出来

**Review by hqkate — APPROVED** [review]

File: configs/rec/robustscanner/README.md [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+|    coco_text     |     42142      |    20    |   real   |
+|      IIIT5K      |      2000      |    20    |   real   |
+|    SynthText     |    2400000     |    1     |  synth   |
+|     SynthAdd     |    1216889     |    1     |    synth    |
+|      Syn90k      |    2400000     |    1     |    synth    |
+
+</div>
+Note: SynthText and Syn90k datasets were both randomly sampled the complete SynthText dataset and complete Syn90k dataset,  consisting of 2.4 million samples each.
+
+The training and evaluation datasets in LMDB format shown in the table above can be downloaded from [here](https://download.mindspore.cn/dataset/robustscanner_dataset/). The downloaded file contains several compressed files, including:
```
Comment (SamitHuang) [inline configs/rec/robustscanner/README.md:90]: datasets uploaded to download server is a concern.

Comment (tonytonglt) [reply configs/rec/robustscanner/README.md]: Fixed, refer to AI Studio download url.

**Review by SamitHuang — APPROVED** [review]
TODO: 1. update model list and what's new in readme. 
2. Add it to online inference.
